mongodb/archivetracking.go

```diff
@@ -0,0 +1,165 @@
+package mongodb
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/ipfs/go-cid"
+	"github.com/textileio/go-threads/core/thread"
+	"github.com/textileio/powergate/ffs"
+	"go.mongodb.org/mongo-driver/bson"
+	"go.mongodb.org/mongo-driver/mongo"
+	"go.mongodb.org/mongo-driver/mongo/options"
+)
+
+type TrackedArchive struct {
+	JID        ffs.JobID
+	DbID       thread.ID
+	DbToken    thread.Token
+	BucketKey  string
+	BucketRoot cid.Cid
+	ReadyAt    time.Time
+	Cause      string
+	Active     bool
+}
+
+// trackedArchive is an internal representation for storage.
+// Any field modifications should be reflected in the cast() func.
+type trackedArchive struct {
+	JID        ffs.JobID    `bson:"_id"`
+	DbID       thread.ID    `bson:"db_id"`
+	DbToken    thread.Token `bson:"db_token"`
+	BucketKey  string       `bson:"bucket_key"`
+	BucketRoot []byte       `bson:"bucket_root"`
+	ReadyAt    time.Time    `bson:"ready_at"`
+	Cause      string       `bson:"cause"`
+	Active     bool         `bson:"active"`
+}
+
+type ArchiveTracking struct {
+	col *mongo.Collection
+}
+
+func NewArchiveTracking(ctx context.Context, db *mongo.Database) (*ArchiveTracking, error) {
+	s := &ArchiveTracking{
+		col: db.Collection("archivetrackings"),
+	}
+	return s, nil
+}
+
+func (at *ArchiveTracking) Create(ctx context.Context, dbID thread.ID, dbToken thread.Token, bucketKey string, jid ffs.JobID, bucketRoot cid.Cid) error {
+	newTA := trackedArchive{
+		JID:        jid,
+		DbID:       dbID,
+		DbToken:    dbToken,
+		BucketKey:  bucketKey,
+		BucketRoot: bucketRoot.Bytes(),
+		ReadyAt:    time.Now(),
+		Cause:      "",
+		Active:     true,
+	}
+	_, err := at.col.InsertOne(ctx, newTA)
+	if err != nil {
+		return fmt.Errorf("inserting in collection: %s", err)
+	}
+	return nil
+}
+
+func (at *ArchiveTracking) GetReadyToCheck(ctx context.Context, n int64) ([]*TrackedArchive, error) {
+	opts := options.Find()
+	opts.SetLimit(n)
+	filter := bson.M{"ready_at": bson.M{"$lte": time.Now()}, "active": true}
+	cursor, err := at.col.Find(ctx, filter, opts)
+	if err != nil {
+		return nil, fmt.Errorf("querying ready tracked archives: %s", err)
+	}
+	defer cursor.Close(ctx)
+	var tas []*trackedArchive
+	for cursor.Next(ctx) {
+		var ta trackedArchive
+		if err := cursor.Decode(&ta); err != nil {
+			return nil, err
+		}
+		tas = append(tas, &ta)
+	}
+	if err := cursor.Err(); err != nil {
+		return nil, err
+	}
+	return castSlice(tas)
+}
+
+func (at *ArchiveTracking) Get(ctx context.Context, jid ffs.JobID) (*TrackedArchive, error) {
+	filter := bson.M{"_id": jid}
+	res := at.col.FindOne(ctx, filter)
+	if res.Err() != nil {
+		return nil, fmt.Errorf("getting tracked archive: %s", res.Err())
+	}
+	var ta trackedArchive
+	if err := res.Decode(&ta); err != nil {
+		return nil, err
+	}
+	return cast(&ta)
+}
+
+func (at *ArchiveTracking) Finalize(ctx context.Context, jid ffs.JobID, cause string) error {
```

Comment:
Do we ever delete the archives tracking objects in mongo? Maybe they should be deleted when the bucket gets deleted?

Reply:
Oh, yeah forgot to mention that. Yeah, now all history gets preserved mostly to keep as much info as possible since all these features are quite new. 
 
What I had in mind was some background job that prunes all `active = false` older than X.  I didn't think about deleting them on bucket deletion, most prob because I'm tilted towards saving as much info as possible in the initial phase.

For sure we won't let this be a problem since is an easy thing to delete. I created #303 to keep track with some of the discussed ideas.

Reply:
Oops, cliked too fast. What do you think @sanderpick ?

Reply:
Nice, SGTM!